Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp

```diff
@@ -0,0 +1,165 @@
+        m_detailHeightClipmapsIndex = m_shaderResourceGroup->FindShaderInputImageIndex(AZ::Name(PassSrgInputs::DetailHeightClipmaps));
+        AZ_Error(TerrainClipmapComputePassName, m_detailHeightClipmapsIndex.IsValid(), "Failed to find terrain srg input image %s.", PassSrgInputs::DetailHeightClipmaps);
+        m_detailMiscClipmapsIndex = m_shaderResourceGroup->FindShaderInputImageIndex(AZ::Name(PassSrgInputs::DetailMiscClipmaps));
+        AZ_Error(TerrainClipmapComputePassName, m_detailMiscClipmapsIndex.IsValid(), "Failed to find terrain srg input image %s.", PassSrgInputs::DetailMiscClipmaps);
+    }
+
+    void TerrainClipmapGenerationPass::FrameBeginInternal(FramePrepareParams params)
+    {
+        ComputePass::FrameBeginInternal(params);
+    }
+
+    void TerrainClipmapGenerationPass::CompileResources(const AZ::RHI::FrameGraphCompileContext& context)
+    {
+        AZ::RPI::Scene* scene = m_pipeline->GetScene();
+        TerrainFeatureProcessor* terrainFeatureProcessor = scene->GetFeatureProcessor<TerrainFeatureProcessor>();
+        if (terrainFeatureProcessor)
+        {
+            auto terrainSrg = terrainFeatureProcessor->GetTerrainShaderResourceGroup();
+            if (terrainSrg)
+            {
+                BindSrg(terrainFeatureProcessor->GetTerrainShaderResourceGroup()->GetRHIShaderResourceGroup());
+            }
+
+            auto material = terrainFeatureProcessor->GetMaterial();
+            if (material)
+            {
+                BindSrg(material->GetRHIShaderResourceGroup());
+            }
+
+            if (m_needsUpdate)
+            {
+                const TerrainClipmapManager& clipmapManager = terrainFeatureProcessor->GetClipmapManager();
+
+                auto macroColorClipmap = clipmapManager.GetClipmapImage(AZ::Name(MacroColorClipmapString));
+                auto macroNormalClipmap = clipmapManager.GetClipmapImage(AZ::Name(MacroNormalClipmapString));
+                auto detailColorClipmap = clipmapManager.GetClipmapImage(AZ::Name(DetailColorClipmapString));
+                auto detailNormalClipmap = clipmapManager.GetClipmapImage(AZ::Name(DetailNormalClipmapString));
+                auto detailHeightClipmap = clipmapManager.GetClipmapImage(AZ::Name(DetailHeightClipmapString));
+                auto detailMiscClipmap = clipmapManager.GetClipmapImage(AZ::Name(DetailMiscClipmapString));
+
+                m_shaderResourceGroup->SetImage(m_macroColorClipmapsIndex, macroColorClipmap);
+                m_shaderResourceGroup->SetImage(m_macroNormalClipmapsIndex, macroNormalClipmap);
+                m_shaderResourceGroup->SetImage(m_detailColorClipmapsIndex, detailColorClipmap);
+                m_shaderResourceGroup->SetImage(m_detailNormalClipmapsIndex, detailNormalClipmap);
+                m_shaderResourceGroup->SetImage(m_detailHeightClipmapsIndex, detailHeightClipmap);
+                m_shaderResourceGroup->SetImage(m_detailMiscClipmapsIndex, detailMiscClipmap);
```

Comment:
You may also consider move this to TerrianFP.

Reply:
This is the pass srg, so FP will have to find and cache the pass. It should be better to do it in the pass itself.

Reply:
It's not necessary to cache pass since you have customized pass here. 

You can replace this block with terrainClipmapManager->SetClipmaptoPassSrg(m_shaderResourceGroup).

I am okay with both. YOu can leave the setImage calls here.

Reply:
I talked to Vicky the other day. It can, but there will be some performance issue about RW and RO access. We don't need to write access in the forward pass, so they are defined as Texture2DArray instead of RWTexture2DArray in terrain SRG.